### ISSUE TITLE
Processing of UBNormal datasets

### ISSUE THREAD

**Issue by 1s3ss** [ristea/aed-mae#6]
Hello, the UBNormal dataset I downloaded is a video file, do I need to extract it as a frame, and how to extract it? And what is the structure of the UBNormal folder?Thanks！

**Comment by Xandra298** [comment]
Same question. The script used to create anomlies needs the frames file of UBNormal, and from the script we can see the dataset has the folder of train/test. However, the original dataset downloaded from official don't has that structure and is the format of videos. Could you please show us some information about that? Is the UBNormal dataset used prepocessed by yourself? (Does this follows the datasplits of the official and all frames of the videos is used?)
Thanks!

**Comment by flww213** [comment]
> 您好，我下载的UBNormal数据集是一个视频文件，我需要将其提取为帧吗，如何提取？UBNormal 文件夹的结构是什么？谢谢！
According to my replication process, simply reading and saving each frame of the MP4 video file is sufficient for success. The dataset format and the extraction code I used are as follows. Additionally, I understand that the ubnormal dataset is only used to generate training data and does not include a test set.

Hope this helps you! (๑•̀ㅂ•́)و✧

-train
--frames
---abnormal_scene_1_scenario_1
----00001.jpg (frames from *.mp4)
----....
---normal_scene_1_scenario_1
--- ....

--annotations
---abnormal_scene_1_scenario_1
----abnormal_scene_1_scenario_1_0000_gt.png 
----abnormal_scene_1_scenario_1_tracks.txt
----...
---normal_scene_1_scenario_1
--- ....

```
import cv2
import os
from tqdm import tqdm 
def save_frames(video_path, output_folder):
    cap = cv2.VideoCapture(video_path)
    if not cap.isOpened():
        print(f"Error opening video file {video_path}")
        return

    os.makedirs(output_folder, exist_ok=True)

    frame_num = 0
    while True:
        ret, frame = cap.read()
        if not ret:
            break

        frame_filename = os.path.join(output_folder, f"frame_{frame_num:04d}.jpg")
        cv2.imwrite(frame_filename, frame)
        frame_num += 1

    cap.release()
    print(f"Frames saved for video {video_path}")

def process_videos(input_folder, output_folder):
    os.makedirs(output_folder, exist_ok=True)
    
    if os.path.isdir(input_folder):
        for filename in tqdm(os.listdir(input_folder)):
            if filename.endswith(".mp4"):
                print(filename)
                video_path = os.path.join(input_folder, filename)
                video_output_folder = os.path.join(output_folder, os.path.splitext(filename)[0])
                save_frames(video_path, video_output_folder)

if __name__ == "__main__":
    indir = "/homeVAD/dataset"  
    outdir = "/home//VAD/datasetframes"  
    process_videos(indir, outdir)
```

**Comment by houchenfeng** [comment]
> Hello, the UBNormal dataset I downloaded is a video file, do I need to extract it as a frame, and how to extract it? And what is the structure of the UBNormal folder?Thanks！

According to my replication process, simply reading and saving each frame of the MP4 video file is sufficient for success. The dataset format and the extraction code I used are as follows. Additionally, I understand that the ubnormal dataset is only used to generate training data and does not include a test set.

Hope this helps you! (๑•̀ㅂ•́)و✧

-train
--frames
---abnormal_scene_1_scenario_1
----00001.jpg (frames from *.mp4)
----....
---normal_scene_1_scenario_1
--- ....

--annotations
---abnormal_scene_1_scenario_1
----abnormal_scene_1_scenario_1_0000_gt.png 
----abnormal_scene_1_scenario_1_tracks.txt
----...
---normal_scene_1_scenario_1
--- ....

```
import cv2
import os
from tqdm import tqdm 
def save_frames(video_path, output_folder):
    cap = cv2.VideoCapture(video_path)
    if not cap.isOpened():
        print(f"Error opening video file {video_path}")
        return

    os.makedirs(output_folder, exist_ok=True)

    frame_num = 0
    while True:
        ret, frame = cap.read()
        if not ret:
            break

        frame_filename = os.path.join(output_folder, f"frame_{frame_num:04d}.jpg")
        cv2.imwrite(frame_filename, frame)
        frame_num += 1

    cap.release()
    print(f"Frames saved for video {video_path}")

def process_videos(input_folder, output_folder):
    os.makedirs(output_folder, exist_ok=True)
    
    if os.path.isdir(input_folder):
        for filename in tqdm(os.listdir(input_folder)):
            if filename.endswith(".mp4"):
                print(filename)
                video_path = os.path.join(input_folder, filename)
                video_output_folder = os.path.join(output_folder, os.path.splitext(filename)[0])
                save_frames(video_path, video_output_folder)

if __name__ == "__main__":
    indir = "/homeVAD/dataset"  
    outdir = "/home//VAD/datasetframes"  
    process_videos(indir, outdir)
```

**Comment by 1s3ss** [comment]
> > 您好，我下载的UBNormal数据集是一个视频文件，我需要将其提取为帧吗，如何提取？UBNormal 文件夹的结构是什么？谢谢！
> 
> 根据我的复制过程，只需读取和保存 MP4 视频文件的每一帧就足以取得成功。我使用的数据集格式和提取代码如下。此外，我了解 ubnormal 数据集仅用于生成训练数据，不包括测试集。
> 
> 希望这对你有所帮助！(๑•̀ㅂ•́)و✧
> 
> -train --frames ---abnormal_scene_1_scenario_1 ----00001.jpg （frames from *.mp4） ----.... ---normal_scene_1_scenario_1--- ....
> 
> --注释 ---abnormal_scene_1_scenario_1 ----abnormal_scene_1_scenario_1_0000_gt.png ----abnormal_scene_1_scenario_1_tracks.txt ----... ---normal_scene_1_scenario_1--- ....
> 
> ```
> import cv2
> import os
> from tqdm import tqdm 
> def save_frames(video_path, output_folder):
>     cap = cv2.VideoCapture(video_path)
>     if not cap.isOpened():
>         print(f"Error opening video file {video_path}")
>         return
> 
>     os.makedirs(output_folder, exist_ok=True)
> 
>     frame_num = 0
>     while True:
>         ret, frame = cap.read()
>         if not ret:
>             break
> 
>         frame_filename = os.path.join(output_folder, f"frame_{frame_num:04d}.jpg")
>         cv2.imwrite(frame_filename, frame)
>         frame_num += 1
> 
>     cap.release()
>     print(f"Frames saved for video {video_path}")
> 
> def process_videos(input_folder, output_folder):
>     os.makedirs(output_folder, exist_ok=True)
>     
>     if os.path.isdir(input_folder):
>         for filename in tqdm(os.listdir(input_folder)):
>             if filename.endswith(".mp4"):
>                 print(filename)
>                 video_path = os.path.join(input_folder, filename)
>                 video_output_folder = os.path.join(output_folder, os.path.splitext(filename)[0])
>                 save_frames(video_path, video_output_folder)
> 
> if __name__ == "__main__":
>     indir = "/homeVAD/dataset"  
>     outdir = "/home//VAD/datasetframes"  
>     process_videos(indir, outdir)
> ```

Ok, I'll try, thank you for your reply！

**Comment by 1s3ss** [comment]
> > Hello, the UBNormal dataset I downloaded is a video file, do I need to extract it as a frame, and how to extract it? And what is the structure of the UBNormal folder?Thanks！
> 
> According to my replication process, simply reading and saving each frame of the MP4 video file is sufficient for success. The dataset format and the extraction code I used are as follows. Additionally, I understand that the ubnormal dataset is only used to generate training data and does not include a test set.
> 
> Hope this helps you! (๑•̀ㅂ•́)و✧
> 
> -train --frames ---abnormal_scene_1_scenario_1 ----00001.jpg (frames from *.mp4) ----.... ---normal_scene_1_scenario_1 --- ....
> 
> --annotations ---abnormal_scene_1_scenario_1 ----abnormal_scene_1_scenario_1_0000_gt.png ----abnormal_scene_1_scenario_1_tracks.txt ----... ---normal_scene_1_scenario_1 --- ....
> 
> ```
> import cv2
> import os
> from tqdm import tqdm 
> def save_frames(video_path, output_folder):
>     cap = cv2.VideoCapture(video_path)
>     if not cap.isOpened():
>         print(f"Error opening video file {video_path}")
>         return
> 
>     os.makedirs(output_folder, exist_ok=True)
> 
>     frame_num = 0
>     while True:
>         ret, frame = cap.read()
>         if not ret:
>             break
> 
>         frame_filename = os.path.join(output_folder, f"frame_{frame_num:04d}.jpg")
>         cv2.imwrite(frame_filename, frame)
>         frame_num += 1
> 
>     cap.release()
>     print(f"Frames saved for video {video_path}")
> 
> def process_videos(input_folder, output_folder):
>     os.makedirs(output_folder, exist_ok=True)
>     
>     if os.path.isdir(input_folder):
>         for filename in tqdm(os.listdir(input_folder)):
>             if filename.endswith(".mp4"):
>                 print(filename)
>                 video_path = os.path.join(input_folder, filename)
>                 video_output_folder = os.path.join(output_folder, os.path.splitext(filename)[0])
>                 save_frames(video_path, video_output_folder)
> 
> if __name__ == "__main__":
>     indir = "/homeVAD/dataset"  
>     outdir = "/home//VAD/datasetframes"  
>     process_videos(indir, outdir)
> ```

Ok, I'll try, thank you for your reply！

**Comment by JPerAsperaadAstra** [comment]
Hi @flww213 , could you share your dataset folder structure after pre-processing? I extracted frames of the avenue dataset and then ran utils/create_anomalies.py to generate abnormal sequences and integrated them into the avenue dataset and got the following dataset structure:
avenue
-training
--frames (extracted original avenue dataset jpg frames)
--frames_abnormal (frames that contains abnormal sequences from the UBnoraml dataset)
--masks_abnormal (masks that corresponds to abnormal sequences)
--gradients2 (generated by extract_gradient.py)

Could you please tell me whether this is the correct data structure that satisfies author's pre-process? I used above processed dataset and ran main.py got an error said: 
[ WARN:0@7.046] global loadsave.cpp:248 findDecoder imread_('/home/me/code/aed-mae/datasets2/avenue/training/frames_abnormal/13/34.jpg'): can't open/read file: check file path/integrity

**Comment by flww213** [comment]
> 嗨，您能在预处理后分享您的数据集文件夹结构吗？我提取了 avenue 数据集的帧，然后运行 utils/create_anomalies.py 生成异常序列并将它们集成到 avenue 数据集中，得到了以下数据集结构： avenue -training --frames （提取的原始 avenue 数据集 jpg 帧） --frames_abnormal （包含来自 UBnoraml 数据集的异常序列的帧） --masks_abnormal （与异常序列相对应的掩码）--gradients2（由 extract_gradient.py 生成）
> 
> 您能否告诉我这是否是满足作者预处理的正确数据结构？我使用了上面处理的数据集并运行 main.py 出现错误说： [ WARN：0@7.046] global loadsave.cpp：248 findDecoder imread_（'/home/me/code/aed-mae/datasets2/avenue/training/frames_abnormal/13/34.jpg'）：无法打开/读取文件：检查文件路径/完整性

1. I made my own masks folder, which is an all-0 mask corresponding to the name in frames. 2. I notice that your image name is 34.jpg, please check that data/train_dataset.py contains len_frame_no to fill in the 0. My data is calculated to 4 digits, which is "0034.jpg".

**Comment by houchenfeng** [comment]
> Hi @flww213 , could you share your dataset folder structure after pre-processing? I extracted frames of the avenue dataset and then ran utils/create_anomalies.py to generate abnormal sequences and integrated them into the avenue dataset and got the following dataset structure: avenue -training --frames (extracted original avenue dataset jpg frames) --frames_abnormal (frames that contains abnormal sequences from the UBnoraml dataset) --masks_abnormal (masks that corresponds to abnormal sequences) --gradients2 (generated by extract_gradient.py)
> 
> Could you please tell me whether this is the correct data structure that satisfies author's pre-process? I used above processed dataset and ran main.py got an error said: [ WARN:0@7.046] global loadsave.cpp:248 findDecoder imread_('/home/me/code/aed-mae/datasets2/avenue/training/frames_abnormal/13/34.jpg'): can't open/read file: check file path/integrity

1. I made my own masks folder, which is an all-0 mask corresponding to the name in frames. 
2. I notice that your image name is 34.jpg, please check that data/train_dataset.py contains len_frame_no to fill in the 0. My data is calculated to 4 digits, which is "0034.jpg".

**Comment by JPerAsperaadAstra** [comment]
@flww213 Thank you so much, I checked the name of the dataset and that's exactly what the problem was!

**Comment by xgh1996** [comment]
@houchenfeng 你好,请问能否加个联系方式,我想咨询下这个项目的问题
